File: android/app/src/main/java/hu/lakat/app/core/Blocklist.kt
package hu.lakat.app.core

/**
 * Domain normalization and preset expansion.
 * Mirrors desktop/src/shared/blocklist.ts — keep the two in sync.
 */
object Blocklist {

    private val DOMAIN_RE =
        Regex("^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?(\\.[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?)+$")

    val PRESETS: Map<String, List<String>> = mapOf(
        "youtube.com" to listOf(
            "m.youtube.com", "music.youtube.com", "youtubei.googleapis.com",
            "youtube-nocookie.com", "www.youtube-nocookie.com", "youtu.be",
        ),
        "facebook.com" to listOf("m.facebook.com", "mbasic.facebook.com", "fb.com", "www.fb.com", "fb.watch"),
        "instagram.com" to listOf("m.instagram.com", "ig.me"),
        "tiktok.com" to listOf("m.tiktok.com", "vm.tiktok.com", "vt.tiktok.com"),
        "x.com" to listOf("twitter.com", "www.twitter.com", "mobile.twitter.com", "mobile.x.com", "t.co"),
        "twitter.com" to listOf("x.com", "www.x.com", "mobile.twitter.com", "mobile.x.com", "t.co"),
        "reddit.com" to listOf("old.reddit.com", "new.reddit.com", "np.reddit.com", "m.reddit.com", "redd.it"),
        "twitch.tv" to listOf("m.twitch.tv"),
        "netflix.com" to listOf("m.netflix.com"),
        "9gag.com" to listOf("m.9gag.com"),
    )

    fun normalizeDomain(input: String): String? {
        var s = input.trim().lowercase()
        if (s.isEmpty()) return null
        s = s.replace(Regex("^[a-z][a-z0-9+.-]*://"), "")
        s = s.replace(Regex("^[^/@]*@"), "")
        val slash = s.indexOfFirst { it == '/' || it == '?' || it == '#' }
        if (slash >= 0) s = s.substring(0, slash)
        val colon = s.indexOf(':')
        if (colon >= 0) s = s.substring(0, colon)
        s = s.trimEnd('.')
        if (s.startsWith("www.")) s = s.removePrefix("www.")
        return if (DOMAIN_RE.matches(s)) s else null
    }

    fun expandHostnames(domain: String, usePreset: Boolean): List<String> {
        val set = sortedSetOf(domain, "www.$domain", "m.$domain")
        if (usePreset) PRESETS[domain]?.let { set.addAll(it) }
        return set.toList()
    }

    /** true when [qname] equals a blocked hostname or is a subdomain of one */
    fun matches(qname: String, blocked: Collection<String>): Boolean {
        for (b in blocked) {
            if (qname == b || qname.endsWith(".$b")) return true
        }
        return false
    }
}

File: ios/Shared/Blocklist.swift
import Foundation

/// Domain normalization + preset expansion.
/// Mirrors desktop/src/shared/blocklist.ts and android .../core/Blocklist.kt.
enum Blocklist {

    private static let domainRegex = try! NSRegularExpression(
        pattern: "^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?(\\.[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?)+$"
    )

    static let presets: [String: [String]] = [
        "youtube.com": ["m.youtube.com", "music.youtube.com", "youtubei.googleapis.com",
                        "youtube-nocookie.com", "www.youtube-nocookie.com", "youtu.be"],
        "facebook.com": ["m.facebook.com", "mbasic.facebook.com", "fb.com", "www.fb.com", "fb.watch"],
        "instagram.com": ["m.instagram.com", "ig.me"],
        "tiktok.com": ["m.tiktok.com", "vm.tiktok.com", "vt.tiktok.com"],
        "x.com": ["twitter.com", "www.twitter.com", "mobile.twitter.com", "mobile.x.com", "t.co"],
        "twitter.com": ["x.com", "www.x.com", "mobile.twitter.com", "mobile.x.com", "t.co"],
        "reddit.com": ["old.reddit.com", "new.reddit.com", "np.reddit.com", "m.reddit.com", "redd.it"],
        "twitch.tv": ["m.twitch.tv"],
        "netflix.com": ["m.netflix.com"],
        "9gag.com": ["m.9gag.com"],
    ]

    static func normalizeDomain(_ input: String) -> String? {
        var s = input.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if s.isEmpty { return nil }
        s = s.replacingOccurrences(of: "^[a-z][a-z0-9+.-]*://", with: "", options: .regularExpression)
        s = s.replacingOccurrences(of: "^[^/@]*@", with: "", options: .regularExpression)
        if let idx = s.firstIndex(where: { $0 == "/" || $0 == "?" || $0 == "#" }) {
            s = String(s[..<idx])
        }
        if let colon = s.firstIndex(of: ":") {
            s = String(s[..<colon])
        }
        while s.hasSuffix(".") { s.removeLast() }
        if s.hasPrefix("www.") { s = String(s.dropFirst(4)) }
        let range = NSRange(s.startIndex..<s.endIndex, in: s)
        return domainRegex.firstMatch(in: s, range: range) != nil ? s : nil
    }

    static func expandHostnames(_ domain: String, usePreset: Bool) -> [String] {
        var set: Set<String> = [domain, "www.\(domain)", "m.\(domain)"]
        if usePreset, let extra = presets[domain] { set.formUnion(extra) }
        return set.sorted()
    }

    /// true when `qname` equals a blocked hostname or is a subdomain of one
    static func matches(_ qname: String, blocked: Set<String>) -> Bool {
        for b in blocked {
            if qname == b || qname.hasSuffix(".\(b)") { return true }
        }
        return false
    }
}
